components/feature/p2p/src/main/java/mozilla/components/feature/p2p/internal/P2PController.kt

```diff
@@ -0,0 +1,232 @@
+                is ConnectionState.Failure -> view.initializeButtons(false, false)
+            }
+        }
+        nearbyConnection?.register(observer, view as P2PBar)
+    }
+
+    fun stop() {
+        nearbyConnection?.unregisterObservers()
+    }
+
+    @Synchronized
+    private fun reportError(msg: String) {
+        Logger.error(msg)
+        view.failure(msg)
+    }
+
+    private inline fun <reified T : ConnectionState> cast() =
+        (savedConnectionState as? T).also {
+            if (it == null) {
+                reportError("savedConnection was expected to be type ${T::class} but is $savedConnectionState")
+            }
+        }
+
+    // P2PView.Listener implementation
+
+    override fun onAdvertise() {
+        nearbyConnection?.startAdvertising()
+    }
+
+    override fun onDiscover() {
+        nearbyConnection?.startDiscovering()
+    }
+
+    override fun onAccept(token: String) {
+        cast<ConnectionState.Authenticating>()?.accept()
+    }
+
+    override fun onReject(token: String) {
+        cast<ConnectionState.Authenticating>()?.reject()
+    }
+
+    override fun onSetUrl(url: String, newTab: Boolean) {
+        if (newTab) {
+            tabsUseCases.addTab(url)
+        } else {
+            sessionUseCases.loadUrl(url)
+        }
+    }
+
+    override fun onReset() {
+        nearbyConnection?.disconnect()
+    }
+
+    override fun onSendPage() {
+        // Because packaging a page takes a long time, we have a status string indicating that
+        // the page is being packaged up. Unlike the other calls to view.updateStatus(), this
+        // does not correspond to a change in NearbyConnection.ConnectionState.
+        view.updateStatus(R.string.mozac_feature_p2p_packaging)
+        if (cast<ConnectionState.ReadyToSend>() != null) {
+            sender.requestHtml()
+        }
+    }
+
+    private fun sendMessage(indicator: Char, message: String) {
+        if (cast<ConnectionState.ReadyToSend>() != null) {
+            when (val messageId = nearbyConnection?.sendMessage("$indicator$message")) {
+                null -> reportError("Unable to send message: sendMessage() returns null")
+                else -> outgoingMessages.put(messageId, indicator)
+            }
+        }
+    }
+
+    fun onPageReadyToSend(page: String) {
+        sendMessage(HTML_INDICATOR, page)
+    }
+
+    override fun onSendUrl() {
+        store.state.selectedTab?.content?.url?.let {
+            sendMessage(URL_INDICATOR, it)
+        }
+    }
+
+    @Suppress("MaxLineLength")
+    override fun onLoadData(context: Context, data: String, newTab: Boolean) {
+        // Store data in file to work around Mozilla bug 1598481, which makes
+        // loading a page from memory extremely inefficient. Instead, we will
+        // write it to a file and then load a URL referencing that file.
+        val file = File.createTempFile("moz", ".html")
```

Comment:
Probably need to do this IO in a coroutine instead of on the main thread.

Reply:
Of course. I'm pleased to get a chance to use a coroutine. They're much nicer than `AsyncTask`.